### PR TITLE
Added padding id option

### PR DESCRIPTION
**Pull request by MarcusLoppe**
Added the option for pad id since not all batches are padded with 0's.

### CODE REVIEW

**Comment by lucidrains** [comment]
@MarcusLoppe oh, so this is actually a decision i made to save on having users having to worry about a mask (pad everything by 0 and then just assuming all 0's on last dimension meant being masked out)

https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/t5.py#L150
https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/open_clip.py#L101

where are you seeing `-1` from?

**Comment by lucidrains** [comment]
@MarcusLoppe could you point me to the `masked_mean` you are referring to?

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe oh, so this is actually a decision i made to save on having users having to worry about a mask (pad everything by 0 and then just assuming all 0's on last dimension meant being masked out)
> 
> https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/t5.py#L150 https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/open_clip.py#L101
> 
> where are you seeing `-1` from?
>

There are two times padding occurs.
1. While the text embeddings are created by the embeddings model (usually padding with 0)
2. Then while training in batches, for each batch it contains a padded version of the text embeddings with a custom pad id from the user.

So for MeshGPT, the custom_collate will pad all the data with -1 instead of the 0 which the embeddings models usually have as padding.

In MeshGPT case, each time the masked_mean was called using the batch which had been padded it had a different mean value since it used the -1 that was in the text embedding.
Masked_mean relied on the mask from TextEmbeddingReturner but since it included the padded dims the mean was always different.

This issue is for all the Conditioners so I probably should have included those in the pull-request.
```
def custom_collate(data, pad_id = -1):
    is_dict = isinstance(first(data), dict)

    if is_dict:
        keys = first(data).keys()
        data = [d.values() for d in data]

    output = []

    for datum in zip(*data):
        if is_tensor(first(datum)):
            datum = pad_sequence(datum, batch_first = True, padding_value = pad_id)
        else:
            datum = list(datum)

        output.append(datum)

    output = tuple(output)

    if is_dict:
        output = dict(zip(keys, output))

    return output
```

**Comment by lucidrains** [comment]
oh fudge! thank you Marcus, will fix the collator

**Comment by MarcusLoppe** [comment]
> oh fudge! thank you Marcus, will fix the collator

I don't think that will work since the autoencoder uses 0 as a valid token. Hence why -1 is required.

**Comment by lucidrains** [comment]
yea you could be right, will look at it this evening

**Comment by MarcusLoppe** [comment]
> yea you could be right, will look at it this evening
>

Awesome, btw little bit off topic but the eps clamp value is applied to the length of the text embedding which is a int/long value.
Is this due to some weird float precision conversion or did you mean to clamp the tensor dims? 
I hope it wasn't to clamp the embedding dim since that would mess up the embedding.

```
def masked_mean(tensor, mask, dim = -1, eps = 1e-5):
    if not exists(mask):
        return tensor.mean(dim = dim)

    mask = rearrange(mask, '... -> ... 1')
    tensor = tensor.masked_fill(~mask, 0.)

    total_el = mask.sum(dim = dim)
    num = tensor.sum(dim = dim)
    den = total_el.float().clamp(min = eps)
    mean = num / den
    mean = mean.masked_fill(total_el == 0, 0.)
    return mean
```

**Comment by lucidrains** [comment]
@MarcusLoppe thank you Marcus! 🙏

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe thank you Marcus! 🙏
>
@lucidrains 
I just noticed I did a blunder with BGE encoder where I didn't include the attention mask and no padding.

Since it would be best to use the same padding for the trainer and the embedding models, should I add the pad id to each model classes and replace the below?:
From:
`encoded_text.masked_fill_(~attn_mask[..., None], 0.)` 
To:
`encoded_text.masked_fill_(~attn_mask[..., None], self.pad_id)`
 
Otherwise you'd need to embed the texts once by one so you don't need the padding.

**Comment by lucidrains** [comment]
@MarcusLoppe i knew you'd pick up on that and already [prepared it](https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/bge.py#L28)! 🙏

**Comment by lucidrains** [comment]
will circle back to this tomorrow morning, bringing the doggo out 🐕

**Comment by MarcusLoppe** [comment]
> @MarcusLoppe i knew you'd pick up on that and already [prepared it](https://github.com/lucidrains/classifier-free-guidance-pytorch/blob/main/classifier_free_guidance_pytorch/bge.py#L28)! 🙏
>

Ah, awesome :) 
I'll get on it.

Another little padding question, in the x-transformers Attention class you can pass the context_mask along with the context, however the k and v values are called with the padded context.
I'm not sure about how well networks can handle/understand that padded dims are not relevant, but would it beneficial to pass the context as mean using something like masked_mean?
Or will the mean on a large scale hurt the model more since the loss of information during the mean?

Attention forward:
```
kv_input = default(context, x)

q_input = x
k_input = kv_input
v_input = kv_input
r_input = x

q = self.to_q(q_input)
k = self.to_k(k_input)
v = self.to_v(v_input) if exists(self.to_v) else k
r = self.to_r(r_input) if exists(self.to_r) else None
```
